File: EaterSpring/src/main/kotlin/xyz/purema/eater/services/ProductOrderService.kt
package xyz.purema.eater.services

import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import xyz.purema.eater.dtos.ProductOrderDto
import xyz.purema.eater.exceptions.NotFoundException
import xyz.purema.eater.interfaces.IMapper
import xyz.purema.eater.interfaces.IProductOrderService
import xyz.purema.eater.repositories.ICustomerRepository
import xyz.purema.eater.repositories.IProductOrderRepository

@Service
class ProductOrderService(
    private val customerRepository: ICustomerRepository,
    private val productOrderRepository: IProductOrderRepository,
    private val mapper: IMapper
) : IProductOrderService {
    override fun getById(id: String): ProductOrderDto {
        val productOrder = productOrderRepository.findById(id).orElseThrow {
            NotFoundException("Order not found")
        }
        return mapper.mapProductOrder(productOrder)
    }

    override fun getAllOrders(): List<ProductOrderDto> {
        val orders = productOrderRepository.findAll(Sort.by(Sort.Direction.DESC, "createdAt"))
        return mapper.mapProductOrders(orders)
    }

    override fun getByCustomerId(customerId: String): List<ProductOrderDto> {
        val customer = customerRepository.findById(customerId).orElseThrow {
            NotFoundException("Customer not found")
        }
        val orders = productOrderRepository.findAllByCustomer(customer)
        return mapper.mapProductOrders(orders)
    }

    override fun getByProductId(productId: String): List<ProductOrderDto> {
        val orders = productOrderRepository.findAllByProductId(productId)
        return mapper.mapProductOrders(orders)
    }
}

File: EaterNet/Eater/Services/ProductOrderService.cs
using Eater.Dtos;
using Eater.Exceptions;
using Eater.Interfaces;
using Mapster;

namespace Eater.Services
{
    public class ProductOrderService : IProductOrderService
    {
        private readonly IProductOrderRepository _productOrderRepository;
        private readonly ICustomerRepository _customerRepository;

        public ProductOrderService(IProductOrderRepository productOrderRepository, ICustomerRepository customerRepository)
        {
            _productOrderRepository = productOrderRepository;
            _customerRepository = customerRepository;
        }

        public async Task<List<ProductOrderDto>> GetAll()
        {
            var orders = await _productOrderRepository.GetAll();
            return orders.Adapt<List<ProductOrderDto>>();
        }

        public async Task<List<ProductOrderDto>> GetAllByCustomerId(string customerId)
        {
            var customer = await _customerRepository.GetById(customerId);
            if (customer == null)
            {
                throw new NotFoundException("Customer not found");
            }

            var orders = await _productOrderRepository.GetAllByCustomer(customer);
            return orders.Adapt<List<ProductOrderDto>>();
        }

        public async Task<List<ProductOrderDto>> GetAllByProductId(string productId)
        {
            var orders = await _productOrderRepository.GetAllByProductId(productId);
            return orders.Adapt<List<ProductOrderDto>>();
        }

        public async ValueTask<ProductOrderDto> GetById(string id)
        {
            var productOrder = await _productOrderRepository.GetById(id);
            if (productOrder == null)
            {
                throw new NotFoundException("Order not found");
            }

            return productOrder.Adapt<ProductOrderDto>();
        }
    }
}
